### PR TITLE
Medical AI - Improve Healing Logic

### PR DESCRIPTION
**Pull request by LachlanSneddon**
**When merged this pull request will:**
-Allows AI to deal with wounded in the cardiac arrest state. 
-Refines AI usage of blood IV's to minimise wastage.

### CODE REVIEW

File: addons/medical_ai/functions/fnc_healingLogic.sqf [hunk 1]
```diff
@@ -54,9 +54,19 @@ switch (true) do {
         _treatmentTime = 5;
         _treatmentArgs = [_target, _selection, "FieldDressing"];
     };
+    case (IN_CRDC_ARRST(_target)): {
+        _treatmentEvent = QEGVAR(medical_treatment,cprLocal);
+        _treatmentTime = 15;
+        _treatmentArgs = [_healer,_target];
```
Comment (mharis001) [inline addons/medical_ai/functions/fnc_healingLogic.sqf]:
```suggestion
        _treatmentArgs = [_healer, _target];    
```

File: addons/medical_ai/functions/fnc_healingLogic.sqf [hunk 2]
```diff
@@ -54,9 +54,19 @@ switch (true) do {
...
     case (_isMedic && {GET_BLOOD_VOLUME(_target) < BLOOD_VOLUME_CLASS_2_HEMORRHAGE}): {
-        private _bloodBags = _target getVariable [QEGVAR(medical,ivBags), []];
-        if ((count _bloodBags) >= 2) exitWith {
+        //check if patient volume + currentIV volume is enough to allow the patient to wake up
+        private _totalIvVolume = 0; //in ml
+        {
+            _x params ["_volumeRemaining"];
+            _totalIvVolume = _totalIvVolume + _volumeRemaining;
+        } forEach (_target getVariable [QEGVAR(medical,ivBags), []]);        
+        if ((GET_BLOOD_VOLUME(_target) + (_totalIvVolume/1000)) > BLOOD_VOLUME_CLASS_2_HEMORRHAGE) exitWith {
```
Comment (mharis001) [inline addons/medical_ai/functions/fnc_healingLogic.sqf]:
```suggestion
        if (GET_BLOOD_VOLUME(_target) + (_totalIvVolume / 1000) > BLOOD_VOLUME_CLASS_2_HEMORRHAGE) exitWith {
```

File: addons/medical_ai/functions/fnc_healingLogic.sqf [hunk 3]
```diff
@@ -54,9 +54,19 @@ switch (true) do {
...
     };
+    case (IN_CRDC_ARRST(_target)): {
+        _treatmentEvent = QEGVAR(medical_treatment,cprLocal);
+        _treatmentTime = 15;
+        _treatmentArgs = [_healer,_target];    
+    };
     case (_isMedic && {GET_BLOOD_VOLUME(_target) < BLOOD_VOLUME_CLASS_2_HEMORRHAGE}): {
-        private _bloodBags = _target getVariable [QEGVAR(medical,ivBags), []];
-        if ((count _bloodBags) >= 2) exitWith {
+        //check if patient volume + currentIV volume is enough to allow the patient to wake up
```
Comment (mharis001) [inline addons/medical_ai/functions/fnc_healingLogic.sqf]:
```suggestion
        // Check if patient's blood volume + remaining IV volume is enough to allow the patient to wake up
```

File: addons/medical_ai/functions/fnc_healingLogic.sqf [hunk 4]
```diff
@@ -54,9 +54,19 @@ switch (true) do {
         _treatmentTime = 5;
         _treatmentArgs = [_target, _selection, "FieldDressing"];
     };
+    case (IN_CRDC_ARRST(_target)): {
+        _treatmentEvent = QEGVAR(medical_treatment,cprLocal);
```
Comment (SilentSpike) [inline addons/medical_ai/functions/fnc_healingLogic.sqf]: What if cpr success chance is set to 0? Then they're just extending the cardiac arrest timer but not actually treating the patient (in such a case the only treatment is a PAK I believe).

Comment (LachlanSneddon) [reply addons/medical_ai/functions/fnc_healingLogic.sqf]: good point, but I believe the PAK needs requires isInStableCondition, so they need to be awake and have a heart rate to use it.

Comment (SilentSpike) [reply addons/medical_ai/functions/fnc_healingLogic.sqf]: I think that requirement might have been removed, but so much has changed I need to double check that 😆

In any case, we should add something to handle this case

Comment (LachlanSneddon) [reply addons/medical_ai/functions/fnc_healingLogic.sqf]: Would it be more prudent to just give them something along the line of:
```
if(GVAR(cprSuccessChance) == 0) exitWith {
_treatmentEvent = QEGVAR(medical_treatment,cprSuccess);
};
```
as a PAK might be a bit overkill on the AI healing system.

Comment (LachlanSneddon) [reply addons/medical_ai/functions/fnc_healingLogic.sqf]: Updated to cover this case. I don't know if it was the right way to go about it but it definitely seems to work.

File: addons/medical_ai/functions/fnc_healingLogic.sqf [hunk 5]
```diff
@@ -54,9 +54,24 @@ switch (true) do {
         _treatmentTime = 5;
         _treatmentArgs = [_target, _selection, "FieldDressing"];
     };
+    case (IN_CRDC_ARRST(_target)): {
+        if(EGVAR(medical_treatment,cprSuccessChance) == 0) then {
```
Comment (mharis001) [inline addons/medical_ai/functions/fnc_healingLogic.sqf]:
```suggestion
        if (EGVAR(medical_treatment,cprSuccessChance) == 0) then {
```

File: addons/medical_ai/functions/fnc_healingLogic.sqf [hunk 6]
```diff
@@ -54,9 +54,24 @@ switch (true) do {
...
+            _treatmentEvent = '#WaitForHeartRate';
```
Comment (mharis001) [inline addons/medical_ai/functions/fnc_healingLogic.sqf]:
```suggestion
            _treatmentEvent = "#WaitForHeartRate";
```

File: addons/medical_ai/functions/fnc_healingLogic.sqf [hunk 7]
```diff
@@ -54,9 +54,24 @@ switch (true) do {
...
+    };
     case (_isMedic && {GET_BLOOD_VOLUME(_target) < BLOOD_VOLUME_CLASS_2_HEMORRHAGE}): {
-        private _bloodBags = _target getVariable [QEGVAR(medical,ivBags), []];
-        if ((count _bloodBags) >= 2) exitWith {
+        // Check if patient's blood volume + remaining IV volume is enough to allow the patient to wake up
+        private _totalIvVolume = 0; //in ml
+        {
+            _x params ["_volumeRemaining"];
+            _totalIvVolume = _totalIvVolume + _volumeRemaining;
+        } forEach (_target getVariable [QEGVAR(medical,ivBags), []]);
```
Comment (mharis001) [inline addons/medical_ai/functions/fnc_healingLogic.sqf]:
```suggestion
        } forEach (_target getVariable [QEGVAR(medical,ivBags), []]);
        
```

File: addons/medical_ai/functions/fnc_healingLogic.sqf [hunk 8]
```diff
@@ -54,9 +54,25 @@ switch (true) do {
         _treatmentTime = 5;
         _treatmentArgs = [_target, _selection, "FieldDressing"];
     };
+    case (IN_CRDC_ARRST(_target)): {
+        if (EGVAR(medical_treatment,cprSuccessChance) == 0) then {
+            _treatmentEvent = "#WaitForHeartRate";
+            [QEGVAR(medical,CPRSucceeded), _target] call CBA_fnc_localEvent;
```
Comment (PabstMirror) [inline addons/medical_ai/functions/fnc_healingLogic.sqf]:
```suggestion
```
I'm not sure about this line, I think if the chance setting is set to 0, then it should fail
The AI should follow the same rules

**Comment by commy2** [comment]
`IN_CRDC_ARRST(_target) && {EGVAR(medical_treatment,cprSuccessChance) > 0}`
as switch condition?

**Comment by PabstMirror** [comment]
I think that makes sense, it should then just use the default of `_treatmentEvent = "#none";`
Death is inevitable, but I can't think of an easy way to make the medic "give up" on the patient

**Pull request merged**